platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,479 @@
+ * limitations under the License.
+ */
+
+package com.swirlds.merkledb.files;
+
+import static com.swirlds.logging.LogMarker.EXCEPTION;
+import static com.swirlds.logging.LogMarker.MERKLE_DB;
+import static com.swirlds.merkledb.files.DataFileCommon.formatSizeBytes;
+import static com.swirlds.merkledb.files.DataFileCommon.getSizeOfFiles;
+import static com.swirlds.merkledb.files.DataFileCommon.getSizeOfFilesByPath;
+import static com.swirlds.merkledb.files.DataFileCommon.logMergeStats;
+
+import com.swirlds.common.config.singleton.ConfigurationHolder;
+import com.swirlds.common.units.UnitConstants;
+import com.swirlds.merkledb.CompactionType;
+import com.swirlds.merkledb.KeyRange;
+import com.swirlds.merkledb.NanoClock;
+import com.swirlds.merkledb.collections.CASableLongIndex;
+import com.swirlds.merkledb.config.MerkleDbConfig;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.UnaryOperator;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This class is responsible performing compaction of data files in a {@link DataFileCollection}.
+ * The compaction is supposed to happen in the background and can be paused and resumed with {@link #pauseCompaction()}
+ * and {@link #resumeCompaction()} to prevent compaction from interfering with snapshots.
+ */
+public class DataFileCompactor {
+
+    private static final Logger logger = LogManager.getLogger(DataFileCompactor.class);
+
+    /**
+     * Since {@code com.swirlds.platform.Browser} populates settings, and it is loaded before any
+     * application classes that might instantiate a data source, the {@link ConfigurationHolder}
+     * holder will have been configured by the time this static initializer runs.
+     */
+    private static final MerkleDbConfig config = ConfigurationHolder.getConfigData(MerkleDbConfig.class);
+
+    private static final Comparator<DataFileReader> DATA_FILE_READER_CREATION_TIME_COMPARATOR =
+            Comparator.comparing(o -> o.getMetadata().getCreationDate());
+    /** Comparator for comparing DataFileReaders by file creation time reversed */
+    private static final Comparator<DataFileReader> DATA_FILE_READER_CREATION_TIME_COMPARATOR_REVERSED =
+            DATA_FILE_READER_CREATION_TIME_COMPARATOR.reversed();
+
+    public static final int DEFAULT_COMPACTION_LEVEL = 0;
+
+    /**
+     * Base name for the data files, allowing more than one DataFileCollection to share a directory
+     */
+    private final String storeName;
```

Comment:
Compactor has a reference to a file collection, which can be used to get the store name. No need to store it here